### ISSUE TITLE
`batch_size_per_gpu` * `grad_accumulation_steps` == global steps ?

### ISSUE THREAD

**Issue by lumpidu** [SWivid/F5-TTS#632]
### Checks

- [X] This template is only for question, not feature requests or bug reports.
- [X] I have thoroughly reviewed the project documentation and read the related paper(s).
- [X] I have searched for existing issues, including closed ones, no similar questions.
- [X] I confirm that I am using English to submit this report in order to facilitate communication.

### Question details

To my understanding, if one likes to reproduce your model with a slightly different hardware, the combination of `batch_size_per_gpu` and `grad_accumulation_steps` should result in the same global batch size and gradient accumulation behaviour.

As an example:

You used 8xA100 80GB GPU's for model training, accordingly your settings are:

```yaml
batch_size_per_gpu: 38400
grad_accumulation_steps: 1
```

If I want to use 8xA100 40GB GPU's for model training to reproduce your results, I'd use:

```yaml
batch_size_per_gpu: 19200
grad_accumulation_steps: 2
```

Then the overall steps should be the same and also training time is roughly the same (same GPU, just half the memory).

But currently, the shown overall steps is not at all dependent from `grad_accumulation_steps`. Also the variable `global_step` in `trainer.py` is always updated independently from the value of `grad_accumulation_steps`.

This is not the case e.g. in [diffusors](https://github.com/huggingface/diffusers/blob/main/examples/unconditional_image_generation/train_unconditional.py#L594).

The consequence is that overall number of steps in the above example is doubled, checkpoints at a certain step-time are therefore not comparable and also training seems to progress twice as fast, as now 2 steps are calculated in the same time where the 80GB A100 version needed 1 step.

It seems to me like a bug, but I am unsure because of your answer in #630.

**Comment by SWivid** [comment]
It is the same updates to compare but not steps (in our paper)

you can find the notes in train.py (previous version)
and also https://github.com/SWivid/F5-TTS/blob/c85252f59a7b2f76bf550140b1306e56508ec6dc/src/f5_tts/configs/F5TTS_Base_train.yaml#L16 the new version using hydra

**Comment by SWivid** [comment]
Sure welcome PR.

The epoch is the number of rounds of going through the dataset, so its concept is natively with steps, but surely someone needs to consider multiplying it according to grad_accum to maintain same update counts.

**Comment by yachuntsaikk** [comment]
I have a similar question.

If I use a single A100 80GB GPU with batch_size_per_gpu: 38400, how can I achieve comparable results to training on 8 x A100 80GB GPUs?

* Should I set grad_accumulation_steps to 8 to simulate the same global batch size?
* Do I also need to scale the training steps (global steps) by 8?
* Or does the number of GPUs only affect training speed without influencing the global batch size?

Thanks!

**Comment by SWivid** [comment]
> I have a similar question.
> 
> If I use a single A100 80GB GPU with batch_size_per_gpu: 38400, how can I achieve comparable results to training on 8 x A100 80GB GPUs?
> 
> * Should I set grad_accumulation_steps to 8 to simulate the same global batch size?
> * Do I also need to scale the training steps (global steps) by 8?
> * Or does the number of GPUs only affect training speed without influencing the global batch size?
> 
> Thanks!

yes, set grad_accum to 8, train to 1.2M*8=9.6M steps (effectively 1.2M updates

**Comment by ndhuynh02** [comment]
does this also mean that if I use a single A100 with 80GB, I can use the batch size of 19200 and 16 grad_accumulation_steps to achieve comparable results
